Added option to select the directory for saving the png file.

rishabhptr opened the pull request:
Fixes #130

rishabhptr commented:
@Ana06 Yes you are right :smile: . Actually it is because when we execute the python scripts we change directory to the output directory so I think if we store the current directory  user would be able to choose path relative to his current path. I have made the change.

Ana06 commented on CHANGELOG.md:
```diff
@@ -5,6 +5,8 @@
 
 ## Version 0.2.0
 
+* Allow to select location to save burndown chart file through `--location`
```
I would say **image** instead of file, to avoid confusion with the `.yml` file.

Ana06 commented on scripts/create_burndown.py:
```diff
@@ -26,12 +26,16 @@ def parseCommandLine():
   args = parser.parse_args()
 
+  if args.location:
+    args.location = os.path.join(os.getcwd(),args.location)
```
what about doing this inside `saveImage`? It is actually not needed at this point. 😉

rishabhptr replied:
Actually, when we pass --output=' ' for indicating the data file we switch to that directory at the very next line of code and saveImage is called in the last so when saveImage is called the current directory would be 'output' directory and not the directory the user is in now so I updated this here. :smile:

Ana06 commented on scripts/plot.py:
```diff
@@ -68,7 +69,15 @@ def drawLegend(self, legend_list):
-    plt.savefig('burndown-' + args.sprint + '.png',bbox_inches='tight')
+    path = args.location
+    filename = 'burndown' + args.sprint + '.png'
+    if path:
```
you could write

```python
if args.location:
  path = os.path.join(os.getcwd(),args.location)
```

rishabhptr commented:
@Ana06 I have added the test but I encountered a issue because as with current functionality we open the image file automatically upon creation which would occur in running test too which isn't desired.

I noticed there is a option '--no-head' for this specific issue but it wasn't working as it was not present in cli.rb so I added it there and using it in my test.Please suggest if I should make some change in this.

Ana06 commented on lib/burndown_chart.rb:
```diff
@@ -160,6 +160,7 @@ def process_options(hash)
         cli_switches << "--output #{hash['output']}" if hash['output']
         cli_switches << '--verbose'                  if hash['verbose']
         cli_switches << "--location #{hash['location']}" if hash['location']
+        cli_switches << '--no-head' if hash['no-head']
```
you shouldn't modify the code to be able to write a test 😱

rishabhptr replied:
Sorry, actually I thought this could be implemented as a option and didnt realize that it had to be passed directly to python scripts for automation.

Ana06 commented on CHANGELOG.md:
```diff
@@ -5,6 +5,8 @@
 
 ## Version 0.2.0
 
+* Allow to select location to save burndown chart image through `--location`
```
this should have been in the _Master (unreleased)_ section, but we have decided in https://github.com/openSUSE/trollolo/pull/163 to stop using this. So please remove this, it is not needed any more. And sorry for the mess. 😄

rishabhptr commented:
@Ana06 Thanks for the review, I have made some changes. I have re-written test to directly use no-head by passing it while executing python script.

Ana06 commented:
@rishabhptr the number of contributors in Trollolo is increasing what is decreasing the velocity to review PR 😞  I'll take a look at soon as possible. 😉

rishabhptr commented:
@Ana06 No problem 😄 , and actually there is more work to it as earlier travis was showing error for not having python-matplotlib installed so I think we might need to add it in travis and also set up some display parameters to successfully execute the python script as directory and file will be created by them.

rishabhptr commented:
@Ana06 The tests are running fine locally, but on travis I think they need matplot installed to run the script so I think a change in travis configuration might be needed to test it properly.

Ana06 commented on .travis.yml:
```diff
@@ -3,6 +3,9 @@ cache: bundler
   - gem --version
   - gem list bundler
+  - sudo apt-get install python-matplotlib
+  - "export DISPLAY=:99.0"
```
why is this needed? :confused:

Ana06 commented on .travis.yml:
```diff
@@ -3,6 +3,9 @@ cache: bundler
   - gem list bundler
+  - sudo apt-get install python-matplotlib
+  - "export DISPLAY=:99.0"
+  - "sh -e /etc/init.d/xvfb start"
```
why is this needed? :confused:

rishabhptr replied:
I had run travis on my fork and had faced issue of not having any display parameter for running python script so I used this here for headless GUI for testing.
https://docs.travis-ci.com/user/gui-and-headless-browsers/

cornelius replied:
Does that mean that you are running matplotlib in the normal unit tests? I would avoid that because the setup can be tricky and strongly depends on the environment. This doesn't give results well defined enough to be reliably checked in tests.

We have the integration tests which run matplotlib in a container: `integration/create_burndown_spec.rb` This works very reliably.

Ana06 replied:
I agree with @cornelius :+1:

rishabhptr replied:
Yes if we have a docker container for matplotlib then it should be used,I will make the changes and also sorry for late response as I was busy with exams past week. :smile:

Ana06 commented on spec/unit/burndown_chart_spec.rb:
```diff
@@ -492,12 +492,21 @@
     end
+
+    it 'saves image file in given location' do
+      dir = given_directory
```
why do you need this variable? :confused:

rishabhptr replied:
It is actually the temporary directory created for writing the test data ( .png file ) ,as  we are using GivenFileSystemSpecHelpers in tests. 
http://www.rubydoc.info/gems/given_filesystem/0.1.2/file/README.md#Thanks

Ana06 replied:
but why do you need to assign it to a variable? :confused:

Ana06 commented on spec/unit/burndown_chart_spec.rb:
```diff
@@ -492,12 +492,21 @@
+      dir = given_directory
+      allow(described_class).to receive(:process_options).and_return(%W{--output spec/data/burndown_dir --location #{dir} --no-head})
+      described_class.plot(2, foo: 1)
+      expect(File.exist?(File.join(dir, 'burndown02.png'))).to be true
```
does this mean that test is creating the Image? Should we maybe delete it? I am not sure if that happens with other tests. :thinking:

rishabhptr replied:
Yes, we are creating the image from our test data that already exists because then only we can test that image would be saved in the right place.
As ,I am mocking the parameters to point '--location' to the temp directory I dont think we need to do anything else as file wont exist after test run.

Also, I did this because this is how we are testing the functionality of '--output' option to  check that data file (.YAML) are saved in correct directory.

https://github.com/openSUSE/trollolo/blob/4f2df0497e6fa6e168eed7cc17d849335312ccda/spec/unit/cli_spec.rb#L14-L20

Ana06 commented:
@rishabhptr what is the state of this? do you need help? :smile:

rishabhptr commented:
@Ana06  As I mentioned earlier in comments I am sorry for late response due to exams. Also if we have docker container for matplotlib it was not required in travis :smile:  so I have made the changes.

rishabhptr commented:
@Ana06 Thanks for the reply :smile: . Sorry for the late response as I was away, I have rebased my branch but I'm encountering a syntax error in one of the python files called by --plot command. I am not really sure if it's same for others.

rishabhptr commented:
@Ana06 Yes the error has now resolved after rebase :smiley: . I have resolved the conflicts, please have a look.

Ana06 commented on scripts/plot.py:
```diff
@@ -68,9 +69,17 @@ def drawLegend(self, legend_list):
   def saveImage(self, args):
-    plt.savefig('burndown-' + args.sprint + '.png',bbox_inches='tight')
+    path = args.location
+    filename = 'burndown-' + args.sprint + '.png'
```
I would name this variable path_image and add the path to it, so that you don't need the else, so:
```ruby
path = args.location
image_path = 'burndown-' + args.sprint + '.png'

if path
  if not os.path.exists(path):
    os.makedirs(path)
  image_path = os.path.join(path, image_path)

plt.savefig(image_path, bbox_inches='tight')

rishabhptr replied:
Yeah this cleans up some code, thanks for the suggestion :smile:

rishabhptr commented:
@Ana06 Thanks for the review, I have made the changes :smile.
